Userland/Applications/SoundPlayer/AudioAlgorithms.cpp

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (c) 2021, Cesar Torres <shortanemoia@protonmail.com>
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "AudioAlgorithms.h"
+#include <math.h>
+
+#include <AK/Format.h>
+
+// This function uses the input vector as output too. therefore, if you wish to
+// leave it intact, pass a copy to this function
+//
+// The sampling frequency must be more than twice the frequency to resolve.
+// The sample window must be at least large enough to reflect the periodicity
+// of the smallest frequency to be resolved.
+//
+// For example, to resolve a 10 KHz and a 2 Hz sine waves we need at least
+// a samplerate of 20 KHz and a window of 0.5 seconds
+//
+// If invert is true, this function computes the inverse discrete fourier transform.
+//
+// The data vector must be a power of 2
+// Adapted from https://cp-algorithms.com/algebra/fft.html
+void fft(complex<double>* data, int sample_count, bool invert)
```

Comment:
I feel like this should be in a more accessible place, since its only use-case isn't audio processing...but maybe it's okay for now?

Either way though, it's a bit sad to have a C-style API for it, maybe use `Span<const complex<double>>`, or just take the vector by const reference?

Reply:
You're right. I put it there because didn't want to extend the scope of the PR too much, but it has much more applications and it'd be worth moving it. I accept suggestions about where to move it :)

Reply:
I think the placement is okay for now, we can move it to a common library when we need it elsewhere.

Reply:
Alright. About the fft using a raw pointer, you're right in your comment. I'll make it more cppish.